lib/active_merchant/billing/gateways/adyen.rb

```diff
@@ -568,16 +568,24 @@ def add_invoice_for_modification(post, money, options)
       end
 
       def add_payment(post, payment, options, action = nil)
-        if payment.is_a?(String)
+        recurring_detail_reference = options[:recurring_detail_reference]
+
+        case payment
+        when String
           _, _, recurring_detail_reference = payment.split('#')
-          post[:selectedRecurringDetailReference] = recurring_detail_reference
           options[:recurring_contract_type] ||= 'RECURRING'
-        elsif payment.is_a?(Check)
+        when Check
           add_bank_account(post, payment, options, action)
         else
-          add_network_tokenization_card(post, payment, options) if payment.is_a?(NetworkTokenizationCreditCard) || options[:wallet_type] == :google_pay
+          add_network_tokenization_card(post, payment, options) if network_tokenization_payment?(payment, options)
           add_card(post, payment)
         end
+
+        post[:selectedRecurringDetailReference] = recurring_detail_reference
```

Comment:
I rather this not be added to this method. It's better to create a separate method for this.

Reply:
Done!